keps/sig-node/3545-improved-multi-numa-alignment/README.md

```diff
@@ -0,0 +1,444 @@
+- [X] Add proper e2e node tests.
+
+#### Alpha to Beta Graduation
+- [X] Gather feedback from the consumer of the `prefer-closest-numa-nodes` policy option.
+- [X] No major bugs reported in the previous cycle.
+
+#### Beta to G.A Graduation
+- [X] Allowing time for feedback (1 year).
+- [X] Risks have been addressed.
+
+
+### Upgrade / Downgrade Strategy
+
+We expect no impact. The new policiy options are opt-in.
+
+### Version Skew Strategy
+
+No changes needed
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `TopologyManagerPolicyOptions`
+  - Components depending on the feature gate: kubelet
+- [x] Change the kubelet configuration to set a `TopologyManager` policy to `restricted` or `best-effort` and a `TopologyManagerPolicyOptions` to `prefer-closest-numa-nodes`
+  - Will enabling / disabling the feature require downtime of the control
+    plane? 
+    No.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node?
+    Yes, a kubelet restart is required.
+
+###### Does enabling the feature change any default behavior?
+
+- Yes, it makes the behaviour of the TopologyManager restricted and best-effort policies to choose NUMA nodes based on distance.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+- Yes, disabling the feature gate shuts down the feature completely.
+- Yes, through kubelet configuration - disable given policy option.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+- No changes. The allocation of resources won't be changed for exiting containers. It will change for new containers only.
+
+###### Are there any tests for feature enablement/disablement?
+
+- There will be specific e2e test demonstrating that default behaviour is preserved when feature gate is disabled.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+Kubelet may fail to start. The kubelet may crash.
+
+###### What specific metrics should inform a rollback?
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+N/A
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No.
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+Inspect the kubelet configuration of the nodes: check feature gate and usage of the new option
+
+###### How can someone using this feature know that it is working for their instance?
+
+In order to verify this feature is working, one should:
+Pick a node with at least 4 NUMA nodes and with at least 3 different distance values between them.
```

Comment:
this is hardly be possible in the current CI environment AFAIK. Are there plans to enable this?

Reply:
I'm not sufficiently familiar with the CI to answer this atm. We could add this as a graduation criteria for beta/GA, and start the discussion in sig-test to see if it's possible to cover this case in e2e tests.

Reply:
That should be sufficient.

Reply:
Can you point at where you added this as a graduation criteria?

I don't see it under the section:
```
**#### Graduation of Options to `Beta-quality` (non-hidden)**
```

Reply:
I have added this as graduation criteria from Beta to GA [here](https://github.com/kubernetes/enhancements/pull/3549/files#diff-71a29ead6cac293d3c3eb9076377f00c3d6726b44ab2e66031e33d4da7dfb08aR316).

Reply:
node e2e tests can choose which kind of machine to run against, e.g. https://github.com/kubernetes/test-infra/blob/master/jobs/e2e_node/image-config-serial-cpu-manager.yaml#L9 used by [pull-kubernetes-node-kubelet-serial-cpu-manager](https://github.com/kubernetes/test-infra/blob/f1fcaa2df0d8e2d01d3e419176b14e98874a49d9/config/jobs/kubernetes/sig-node/sig-node-presubmit.yaml#L459)

but I'm not sure if that will allow you to be as granular as you're describing